### ISSUE TITLE
FR: Full text search: Result info notifications; Allow page turns

### ISSUE THREAD

**Issue by jonnyl2** [koreader/koreader#11203]
Although there are already a number of other Fulltext search-related FRs and discussions open (e.g., most recently #11038) that would address at least the first issue, I would like to share a simpler idea that wouldn't require a complete redesign of the Fulltext search feature, in case devs find this useful as well and someone is interested in adding this (at least for the shorter term, until any larger project is undertaken):

### 1. Notifications with search info

**Does your feature request involve difficulty completing a task? Please describe.**

When using Fulltext search in a book, the UI gives no indication as to the number of results found or whereabouts in the book the results are located. It would be useful to get some idea of how many search results there are, on how many separate pages they are located, and how spread around in the book they are.

**Describe the solution you'd like**

Currently, the following notifications are displayed at the beginning and end of the search results:
`No results on preceding pages` and `No results on following pages` (when attempting to go back or forward, respectively).

How about also displaying notifications on each page with search results, for example like this:

On page 51: `1/5 Displaying results 1-6 of 40 [51-193]`
On page 53: `2/5 Displaying results 7-10 of 40 [51-193]`
On page 58: `3/5 Displaying result 11 of 40 [51-193]`
On page 99: `4/5 Displaying results 12-31 of 40 [51-193]`
On page 193: `5/5 Displaying results 32-40 of 40 [51-193]`

(Where the numbers in square brackets represent the range of pages with any results found.)

### 2. Page turns

**Does your feature request involve difficulty completing a task? Please describe.**

When looking at search results, page turning is not possible. Page turning is sometimes desirable to confirm if the correct spot was found, or just to read more context around the result before moving on to the next result. Turning the page currently requires exiting search and then relaunching it to review further results (and cycle through all previous results again).

**Describe the solution you'd like**

Allow page turning while search results are being displayed. Either by left/right swiping on the page or add extra `-1` `+1` page turn buttons on the ButtonDialog.

Thanks for reading.

**Comment by ryanwwest** [comment]
I think both of these additions would be very helpful and a much smaller stepping stone to improving fulltext search before having a revamped multipage result page.

**Comment by poire-z** [comment]
Sharing some points and thoughts about #11038 and this issue.

Getting all the search results in the book could be slow, and memory hungry (ie. if you search for `e` :)).

I'd rather not have limits or heuristics, or user configurable, like "pattern must be 3 chars min" (because Chinese may want to search for just `爨`) or "max 100 results" (then, which ones? starting from the beginning, the current page, forward or backward?) 
Also, what is context ? A full paragraph, a sentence, a few words after/before (again, the notion of words or needed context may differ by language), and some results may have no context (ie. a country name in a table cell).

Our current behaviour (which has existed since before I joined the project) is incremental and fast, as we just scan in one direction until we find results, and stop there - and restart from there.
It indeed does not tell you anything about how many results, but it gives you the best context possible :)

(As for a full search, I guess it should be as fast or slow as looking for some non existant pattern and reaching the end of book - which feels quite fast though - except for the added accumulation/compilation of results.)

Also, it currently works quite fine with both our modules/engines - implementing a full result list may need some huge works on both sides :/ 

> How about also displaying notifications on each page with search results, for example like this:
`On page 51: 1/5 Displaying results 1-6 of 40 [51-193]`

I know you like numbers soup with bracketnoodles :)
But I'm not certain this gives valuable information - or may be too much.
Moreover, our notifications do time-out (by default, could be annoying if they don't and it needs user action to make them go away and see the text below), so you wouldn't really have the time to pick something out from this long and complex string.

> When using Fulltext search in a book, the UI gives no indication as to the number of results found or _whereabouts in the book the results are located_. It would be useful to get some idea of how many search results there are, _on how many separate pages they are located_, and _how spread around in the book they are_.

I'm surprised (and actually worried :) but also relieved to be honest :)) that you didn't mention "Book map" ! :)
Pretending/assuming we could get all the search results, you could have all the points in italic above if we would show some symbol on page with search results (like for bookmarks, notes, previous location) on the book map. (I had that thought before: https://github.com/koreader/koreader/pull/8613#issuecomment-1006166298.)
Taping near them could bring Page browser, with search results highlighted (this may be hard to implement though), may be forced in 1x1 rows.columns, so one could read text and context, and turn pages, without affecting current page and infecting statistics.

(I remember that in a previous life where I was reading paper books, I usually had a quite visual memory of where I've seen some words or ideas, like "it was in the bottom third of a left page", so when scanning back to find it again, I could quickly glance at just this visual area - most often without success :) Dunno if this would still work on an ereader...)

> Page turning is sometimes desirable to confirm if the correct spot was found, or just to read more context around the result before moving on to the next result. 
> Allow page turning while search results are being displayed. Either by left/right swiping on the page or add extra -1 +1 page turn buttons on the ButtonDialog.

I agree this would be nice to have. With L/R swipe would probably be the least annoying and intrusive (except on Non-Touch devices).
Might not be that easy to implement, if we turn pages and still want new search results to be highlighted, or to properly rescan for results from the last real page or from the turned page...

**Comment by jonnyl2** [comment]
>I know you like numbers soup with bracketnoodles :)

Indeed :)

>But I'm not certain this gives valuable information - or may be too much.
Moreover, our notifications do time-out (by default, could be annoying if they don't and it needs user action to make them go away and see the text below), so you wouldn't really have the time to pick something out from this long and complex string.

I disagree with this. The notification time is quite long, and even if it is too short for some reason to catch all the information displayed, it would still be helpful. The missing part would just be picked up when paging to the next screen. Or the user might only be interested in one part of the notification display anyway (e.g. the number of pages with results, or the interval) and only needs to look at that part of the screen. 

Anyway, even just a notification `4/32` (that shows current results page out of number of all pages with results) would still be a lot more helpful than leaving the user completely in the dark about when they'll reach the end of the results.

>I'm surprised (and actually worried :) but also relieved to be honest :)) that you didn't mention "Book map" ! :)

It _did_ cross my mind, but I wanted to keep this FR as simple as possible.

>Taping near them could bring Page browser, with search results highlighted (this may be hard to implement though), may be forced in 1x1 rows.columns, so one could read text and context, and turn pages, without affecting current page and infecting statistics.

This would certainly be nice and would also make the notifications for the results paging superfluous, and probably meet the expectations of #11038 as well.

>>Allow page turning while search results are being displayed. Either by left/right swiping on the page or add extra -1 +1 page turn buttons on the ButtonDialog.

>I agree this would be nice to have. With L/R swipe would probably be the least annoying and intrusive (except on Non-Touch devices).
Might not be that easy to implement, if we turn pages and still want new search results to be highlighted, or to properly rescan for results from the last real page or from the turned page...

Perhaps it would be easier to implement if the current state is saved, so that when the user taps the screen, they can navigate the document as usual, and when they relaunch Fulltext search, it will go back to where he was? (One could add an extra button `Exit` that would abandon search without saving. Or the opposite approach; a button that will save and close (e.g., `Temp close` or `Reader` or `Browse` or ???) while tapping the page will exit without saving as it does now.

**Comment by ryanwwest** [comment]
Was about to comment after forgetting a few days back and saw that @hius07 already has a change incoming, wow. I'm not sure if a temporary notification with that information would work well for me, just knowing that it might disappear after a few seconds - I'd prefer the content stay visible until I manually dismiss it along with the other search dialog. Though perhaps the new PR will make the non-page-turns part of this FR less necessary, I'm not sure. Especially since we want to keep the regular search outside of the all-results page snappy.

**Comment by jonnyl2** [comment]
Closing as #11313 far exceeds my expectations from this FR. Thank you.

**Comment by poire-z** [comment]
Random hijack on a search related issue (by somebody who knows about the location stack).
Also pinging @yparitcher (who liked that back location stack so much he added a forward location stack).

When doing a full text search from selecting a word(s), it often happens that I get a single result: the one I selected - so `<` or `>` move nowhere else, and I just dismiss the dialog, and go on reading. Currently, just doing that first search adds the current page to the location stack (so, if you go on reading, and swipe back, you may be back there and lose your current position).
Is this desirable ? For consistency (so, one used to doing that can always do that swipe back after a search to clear it)?
Or should we delay adding the current location to the back stack until we move from the current page? (so, a bit of inconsistency).

I could change the behaviour with this:
```diff
--- a/frontend/apps/reader/modules/readersearch.lua
+++ b/frontend/apps/reader/modules/readersearch.lua
@@ -358,7 +358,16 @@ function ReaderSearch:onShowSearchDialog(text, direction, regex, case_insensitiv
                     end
                 end
                 -- Don't add result pages to location ("Go back") stack
-                neglect_current_location = true
+                if not neglect_current_location then
+                    if self.view.view_mode == "page" and current_page == self.ui.document:getCurrentPage() then
+                        -- Initial search, but we didn't change page: remove the location onGotoLink
+                        -- has added, and we will re-add it next time (if) we do change page.
+                        table.remove(self.ui.link.location_stack)
+                        neglect_current_location = false
+                    else
+                        neglect_current_location = true
+                    end
+                end
             end
             if no_results then
                 local notification_text
```
(I'll probably use that in my own patch, unless people things it's worth changing the current behaviour.)

**Comment by Frenzie** [comment]
Does it even make sense to have the same page in there twice in a row? I'd say that sounds more like a bug or at least an unintended side effect in a rare (?) scenario.

**Comment by poire-z** [comment]
> Does it even make sense to have the same page in there twice in a row?

?
In case you mean that it should immediately jump to results in another page than the current - no, it shouldn't do that, as it higlights the possible other matches in the current page - and allows you to go seach backwards if that's your aim.

**Comment by jonnyl2** [comment]
I've never (consciously) encountered this issue but it'd make sense IMO. Unwanted page histories in the location stack can be bothering, yes. 

This actually reminds me of a similar issue with a potentially alternative solution: Offer a way to exit the search dialog without leaving any traces, so that it jumps back to the page where the search was launched, without adding anything to the stack. I often search for a term or character name's first mention to review its definition or their description. I exit the dialog there, and then jump back to the previous location. Now I have those 2 locations in the stack, even thought I have no further use for referring back to those locations. IMO it would be much cleaner if I could just read the definition/description while having the search dialog open, and then exit out of it (with an extra button or multiswipe, for example) and it will close the dialog and jump back to its origin without any location history. Thoughts?

**Comment by poire-z** [comment]
> I exit the dialog there, and then jump back to the previous location. Now I have those 2 locations in the stack

You mean the "forward stack" ? Because your back stack should be empty when you jump back.
(I have no use for that forward stack, and it doesn't bother me. - except when I indeed have an unwanted back stack that I go to :) that "Go forward" indeed helps getting where I was, and then I can just clear the location stack(s)).

> IMO it would be much cleaner if I could just read the definition/description while having the search dialog open, and then exit out of it (with an extra button or multiswipe, for example) and it will close the dialog and jump back to its origin without any location history.

It's simpler to just dismiss that dialog - and allow page turning (to read more of that definition on the next page) - than having additionial button to prev/next pages.
And that "go back" gesture is the ubiquitous way to... "go back" :) it's in the end not more complicated than doing a specific multiswipe.

(And when we have a dialog/widget on top, it's complicated to get any gesture associated to the underneath widget (the Reader) working.)

**Comment by Frenzie** [comment]
> > Does it even make sense to have the same page in there twice in a row?
> 
> ? In case you mean that it should immediately jump to results in another page than the current - no, it shouldn't do that, as it higlights the possible other matches in the current page - and allows you to go seach backwards if that's your aim.

Maybe I misunderstood, but I thought you said it was added to the location stack twice, meaning you'd go "back" only to remain on the same page. There's no inconsistency in not having it in there twice, imho.

**Comment by poire-z** [comment]
No, it is added only once. Early, when the search shows you the matches on the current page, before you move to another location, it adds the current page (that you have not left) to the stack. When you move to other results, nothing more is added to the stack (because the origin page has already been added).

When you do a full text search from the input dialog, for some words that are not on the current page, the current page is added to the stack, and you are brought to some other page.

So it is consistent it is added to the stack (and technically simpler) - it's just unpractical/confusing it is added before you actually move away.

**Comment by jonnyl2** [comment]
>You mean the "forward stack" ? Because your back stack should be empty when you jump back.

Yes, that must be it. I didn't realize they were separate. Is that the reason some locations are not displayed in the Book map? Is it just the back stack that gets numbered dots?

>It's simpler to just dismiss that dialog - and allow page turning (to read more of that definition on the next page) - than having additionial button to prev/next pages.

In this example I meant in cases where I just want to exit the dialog back directly back to the original location without any page turning at the result location at all.

**Comment by poire-z** [comment]
> Is it just the back stack that gets numbered dots?

Yes, I just cared about the back stack. I don't really understand what's that forward stack is about :) (except may be to recover from accidents). It was added in #10228 .

**Comment by Frenzie** [comment]
> So it is consistent it is added to the stack (and technically simpler) - it's just unpractical/confusing it is added before you actually move away.

But is it confusing in code or is it confusing to the user as going "back" with nothing happening the first time?

**Comment by poire-z** [comment]
None.
It's confusing (to me - possibly to nobody else :)) because if there is no further matches on other pages, and the user (me) not seeing we moved pages, won't think about "going back" (to clear the stack of the current page that was added) - and he will carry an unwanted "back location" in the stack (with possible accidental go back, or a little ❶ in Book map - that's how I usually see it).

**Comment by yparitcher** [comment]
@poire-z 
I did not envision that use case when i redid the location stack code, i agree though (now that you mention it) that we should only add it once a page has changed.

The forward stack is for flipping between 2 pages (think of a physical book :) when i want to easily go between 2 parts of a book that cross reference or such.

**Comment by poire-z** [comment]
I ended up with this:
```diff
--- a/frontend/apps/reader/modules/readerlink.lua
+++ b/frontend/apps/reader/modules/readerlink.lua
@@ -714,2 +714,6 @@ end

+function ReaderLink:popFromLocationStack()
+    return table.remove(self.location_stack)
+end
+
 function ReaderLink:onClearLocationStack(show_notification)
--- a/frontend/apps/reader/modules/readersearch.lua
+++ b/frontend/apps/reader/modules/readersearch.lua
@@ -291,2 +291,5 @@ function ReaderSearch:onShowSearchDialog(text, direction, regex, case_insensitiv
                 if self.ui.paging then
+                    if not current_page then -- initial search
+                        current_page = self.ui.paging.current_page
+                    end
                     no_results = false
@@ -359,4 +362,20 @@ function ReaderSearch:onShowSearchDialog(text, direction, regex, case_insensitiv
                 end
-                -- Don't add result pages to location ("Go back") stack
-                neglect_current_location = true
+                if not neglect_current_location then
+                    -- Initial search: onGotoLink() has added the current page to the location stack,
+                    -- and we don't want this to be done when showing further pages with results.
+                    -- But if this initial search is showing results on the current page, we don't want
+                    -- the original page added: we will do it when we jump to a different page.
+                    -- For now, only do this with CreDocument. With PDF, whether in single page mode or
+                    -- in scroll mode, the view can scroll a bit when showing results, and we want to
+                    -- allow "go back" to restore the original viewport.
+                    if self.ui.rolling and self.view.view_mode == "page" then
+                        if current_page == (self.ui.rolling and self.ui.document:getCurrentPage() or self.ui.paging.current_page) then
+                            self.ui.link:popFromLocationStack()
+                            neglect_current_location = false
+                        else
+                            -- We won't add further result pages to the location stack ("Go back").
+                            neglect_current_location = true
+                        end
+                    end
+                end
             end
```
Trying to find out when to do it or not with PDF, and not getting some consistent answer (when searching and showing results on the page, there may be a bit of scrolling, but the initial y-scroll position is sometimes stored/restored, sometimes not), I ended up not doing anything for PDFs.
So, well, the above works for me and my use case, but not super happy with not having a consistent behaviour.
Should I keep that for me, or PR it anyway ?

**Comment by jonnyl2** [comment]
>So, well, the above works for me and my use case, but not super happy with not having a consistent behaviour.
Should I keep that for me, or PR it anyway ?

FWIW, for me it's a welcome change to the code. Tested and worked as described (on a CRE-doc).

No real idea about PDF either. There are other search related issues there. For example, if using Fulltext search when zoomed in on a page, the search result will only show if it's in the first rectangle of a page. If it's outside, it will turn to that page but the result is not visible. Similar with reflow mode: it has to be in the text that fits on the first reader-screen of a book-page, or it will not scroll/jump to it and will not be visible.
